Fix deprecated copy for frames

wxmerkt opened the pull request:
Since we are explicitly declaring the copy constructor for the frames, we need to also explicitly declare the assignment operator. This becomes a warning in more recent versions of gcc/C++ standards.

proyan commented:
Cool. Thanks @wxmerkt!

cmastalli commented on include/crocoddyl/core/activations/quadratic-barrier.hpp:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2019-2020, LAAS-CNRS, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Would you write this in the same line? (Same for the other files)

This is just to simplify the copyright signature. Indeed, I did this simplification a few PR ago.

cmastalli commented on include/crocoddyl/core/optctrl/shooting.hxx:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2019-2020, LAAS-CNRS, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Same here

cmastalli commented on include/crocoddyl/multibody/frames.hpp:
```diff
@@ -3,6 +3,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2019-2020, LAAS-CNRS, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
And here

cmastalli commented on include/crocoddyl/multibody/friction-cone.hpp:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2019-2021, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Here

cmastalli commented on include/crocoddyl/multibody/friction-cone.hxx:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2019-2021, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Here

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hpp:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2020-2021, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Here

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -2,6 +2,7 @@
 // BSD 3-Clause License
 //
 // Copyright (C) 2020-2021, University of Edinburgh
+// Copyright (C) 2021, University of Oxford
```
Here

cmastalli reviewed (changes requested):
Thank you so much @wxmerkt.

I have two main requests:
 1. Could you add unittests for the FrictionCone and WrenchCone in the 'test_constructor' functions? (To check the new operator introduced in this PR)
  2. Could you add this constructor to the CoMSupport class? (Including unittest as in 1)

There is also another minimal request (yet not crucial). It would be nice if you add the copyright in the same line with other copyholders (just to simplify the code).

wxmerkt commented:
Thanks for the feedback, I've addressed the comments and added the unit tests. I've also added code to explicitly ignore deprecation warnings in the Python bindings. The project is now free of warnings.

cmastalli reviewed (approved):
Thank you so much! also for the ignore deprecation macros!

@proyan --- thank you so much for also reviewing the code.

proyan commented:
@wxmerkt The tests are failing,
```
14/42 Test #14: test_diff_actions .............................***Failed    0.94 sec

Running test_DifferentialActionModelLQR

Running test_DifferentialActionModelLQRDriftFree

Running test_DifferentialActionModelFreeFwdDynamics_TalosArm

Running test_DifferentialActionModelFreeFwdDynamics_TalosArm_Squashed

Running test_DifferentialActionModelContactFwdDynamics_TalosArm

Running test_DifferentialActionModelContactFwdDynamics_HyQ

Running test_DifferentialActionModelContactFwdDynamics_Talos

Running test_DifferentialActionModelContactFwdDynamicsWithFriction_TalosArm

Running test_DifferentialActionModelContactFwdDynamicsWithFriction_HyQ

Running test_DifferentialActionModelContactFwdDynamicsWithFriction_Talos

Running 50 test cases...

/home/travis/build/loco-3d/crocoddyl/unittest/test_diff_actions.cpp(127): error in "boost::bind(&test_partial_derivatives_against_numdiff, action_type)": check (data->Fx - data_num_diff->Fx).isMuchSmallerThan(1.0, tol) failed

*** 1 failure detected in test suite "Master Test Suite"
```
Could you please take a look?

jcarpent commented:
You should use isZero instead of isMuchSmallerThan.

cmastalli commented:
@wxmerkt  -- I should probably handle the issue with the unit-tests as this is not affected by the work on this PR.
The only issue is that I could not work now (but asap) on it.

cmastalli commented:
> I am aware that likely there is an issue with my code in this PR - it does not yet handle self-assignment checks for the assignment operator (`if (this != &other)`).

 I'll let you solve this issue. Then, I'll fix any issue in the unit-test (if they are still happening after handling this)

> You should use isZero instead of isMuchSmallerThan.

@jcarpent -- it makes sense, and I happy to handle this in another PR as all the unit-test is written in this way.
Any suggestion regarding the tolerance? using the same one?

wxmerkt commented:
I did some more digging and switched back to gcc - the test failure is already present on `devel`. I added an output of the norm between the Fx and num-diff-Fx as:
```cpp
std::cout << action_type << " - DIFF - " <<  (data->Fx - data_num_diff->Fx).norm() << std::endl;
```
in `test_diff_actions.cpp`, L127.

The output we get is then:
```
DifferentialActionModelContactFwdDynamics_Talos - DIFF - 0.00111288
crocoddyl/unittest/test_diff_actions.cpp(128): error: in "test_DifferentialActionModelContactFwdDynamics_Talos/boost__bind(&test_partial_derivatives_against_numdiff, action_type)": check (data->Fx - data_num_diff->Fx).isMuchSmallerThan(1.0, tol) has failed
```

I'll get on fixing the self-assignment issue in the PR and update it with that.

jcarpent commented:
The tolerance is `tol = sqrt(eps)` where eps is the increment of the finite diff. Please see Pinocchio for further examples.

wxmerkt commented:
This PR should now be ready:
- I went back to `devel` and made sure I can get the unittests to pass there with gcc 9.3.0 in Debug mode: This required a modification to the tolerance for Fx, Fu in `test_diff_actions`. I based the general tolerance on `sqrt(eps)`, but had to adapt for Fx and Fu.
- I added the self-assignment check for the assignment operator. I also found another missing copy constructor in `CoPSupport`.

jcarpent commented on unittest/test_friction_cone.cpp:
```diff
@@ -52,6 +52,37 @@ void test_constructor() {
+    crocoddyl::FrictionCone cone_reference(cone);
+
+    BOOST_CHECK(cone.get_nf() == cone_reference.get_nf());
+    BOOST_CHECK((cone.get_A() - cone_reference.get_A()).isMuchSmallerThan(1.0, 1e-9));
```
`isMuchSmallerThan` is not very elegant.

wxmerkt commented:
@nim65s Gepgitlab is failing due to "out of harddrive space"

cmastalli commented:
@wxmerkt and @jcarpent  -- I have fixed the unit-test request in #910. However, the contact-dynamics tests do not pass with `sqrt(eps)`. Instead, we need to use a factor of `10`, i.e., `10*sqrt(eps)`.

I suggest to do not work more on the unittests in this PR.

wxmerkt commented:
This PR is complete and can be merged. The unit tests have also been fixed. I suggest to merge this one before #910
